File: 8.回溯/2.组合问题/77.组合.cpp
/*
 * @lc app=leetcode.cn id=77 lang=cpp
 *
 * [77] 组合
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> res;
        vector<int> pack;
        dfs(1, n, k, res, pack);
        return res;
    }
    void dfs(int sta, int n, int k, vector<vector<int>>& res, vector<int>& pack)
    {
        if (pack.size() == k)
        {
            res.emplace_back(pack);
            return;
        }
        for (int i = sta; i <= n - (k - pack.size()) + 1; i ++ )
        {
            pack.push_back(i);
            dfs(i + 1, n, k, res, pack);
            pack.pop_back();
        }
    }
};
// @lc code=end

File: 面经150/15.回溯/77.组合.py
#
# @lc app=leetcode.cn id=77 lang=python3
#
# [77] 组合
#
from typing import List
# @lc code=start
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        res = []
        pack = []
        def dfs(sta):
            if len(pack) == k:
                res.append(pack[:])
                return
            for i in range(sta, n + 1):
                pack.append(i)
                dfs(i + 1)
                pack.pop()
        dfs(1)
        return res
# @lc code=end

